### ISSUE TITLE
Based on reviewer feedback: Clarify that we are NOT claiming learnability only feasibility though we identify all our attention-head compatible grammar pattern matchers with training examples (Table 2)

### ISSUE THREAD

**Issue by willy-b** [willy-b/RASP-for-ReCOGS#71]
**For the separate question of whether proving that a flat/non-hierarchical solution is feasible without addressing learnability is even interesting? Isn't it obvious?**
See https://github.com/willy-b/RASP-for-ReCOGS/issues/70  for that separate question.

**Regarding human construction of our RASP program and whether it implies the approach is not learnable or not**
We are not claiming learnability, only feasibility by Encoder-Decoder Transformers in the paper, though we do take pains to ensure each RASP rule can be trivially derived from at least one specific training example (see Table 2).

The approach (in the paper Results (6), first page and Appendix 9.2, but summarized here):
We use word level tokens, and an embedding that tags each sequence position with possible parts of speech / verb types, after which a flat pattern matching approach based on masking out “pp det common_noun” and “pp proper_noun” and then matching against 19 templates which are trivially derived from training examples (see Table 2 and Appendix 9.11) is applied. The matched template gives the relationships of the verb in the same order as the non-pp nouns in the sentence match to them, so the noun-verb relationships can be read off by the decoder loop in the order given in the sentence.

**How the RASP-for-ReCOGS model was designed:**

In the paper we cite (Tenney et al 2019) “BERT rediscovers the classical NLP pipeline” who find that immediately post-embedding (layer 0) of a Transformer trained with a language modeling objective part of speech information is already tagged. Based on what Transformers are actually observed to do in that paper we also mapped input sentences to their parts of speech. The author expected that this would necessarily be the beginning of a long and extremely complicated pipeline to solve the problem requiring hierarchical transformations as are observed in deep neural network computer vision models (or even e.g. the CKY algorithm for parsing CFG sentences, which first handles length 1 parses, length 2 parses, etc up to the full sequence; and we had (Klinger et al., 2024)’s useful CFG description of the COGS/ReCOGS input grammar on hand which is a hierarchical description (we use grammar coverage to show Klinger's description is covered by the examples we derive flat pattern matches for, and to argue why certain mistakes are indicative of NOT using such a hierarchical representation)), but the author was wrong in this case. After switching to a possible part-of-speech sequence (nit: use 5 parallel sequences so that e.g. the maximally ambiguous verb “liked” can have 4 possible verb types, could be also be 1 sequence in higher dimension instead of 5 parallel scalar sequences), there was one further natural reduction to treat “det common_noun” and “proper_noun” the same everywhere. After applying just these two transformations, one sees that if one first excludes COGS/ReCOGS input sentences with recursive grammar (sentential complements and prepositional phrases), there are actually only 19 unique sentences left in the training data (instead of tens of thousands), which each have a simple mapping in terms of noun order to verb relationship order (recall in ReCOGS logical form one lists the nouns and verbs, then outputs the verb relationships). If one puts back in the preposition sentences one sees that simply masking out “pp det common_noun” and “pp proper_noun” returns one to the same 19 sentences in the entire training set. The tail-recursive “pp common_noun”/”pp proper_noun” information is only used in the “nmod” part of the logical form which can be output by the decoder loop at the end of the logical form generation. Similarly the sentential complements can be masked away when recognizing the main grammar pattern of the sentence which determines all of the non-CP parts of the logical form.
This may be handcrafted, but it is very simple and based on an embedding that real networks have learned from scratch (albeit Tenney et al 2019 is with a language modeling objective instead of seq2seq objective; regarding this we already cite in our paper the intriguing work of (Ahuja et al., 2024; “Learning syntax without planting trees: Understanding hierarchical generalization in transformers”) who provide examples of solving the same language tasks using seq2seq vs various language modeling objectives - and they indeed find better generalization performance on their problems by just making the change from seq2seq to the language modeling objective (training to model both the input and the output)). 
Again we don't claim learnability in this paper but I don't think trivial memorization of 19 flat patterns or the few transformations required to reduce the dataset to the 19 flat patterns require special human ingenuity to do (and implementing in RASP shows it is feasible by the Transformer).
In fact based on the Wu et al 2023 baseline failure on v_dat_p2 recipient pp modification generalization (Figure 3) and commission of attraction errors (Figure 7), I think we have evidence the Wu et al 2023 baseline Encoder-Decoder Transformer is already using a similar flat-pattern matching strategy but without (unfortunately) having learned to mask out / ignore the “pp det common_noun” and “pp proper_noun” attractors everywhere.

**Comment by willy-b** [comment]
Update description here to avoid confusion (this is quite different than internal discussions I was just noting work to be done).

Added here:
We have already noted in our conclusion that we don’t yet know how to get the Transformer trained from scratch to learn the pp masking rules: "We predict that Transformers should be able to perform the ReCOGS task (even structural generalization splits) with high accuracy and that the problem is just of getting the Transformer to learn the appropriate rules" with footnote "e.g. to ignore 'pp det common_noun' and 'pp proper_noun' when finding nouns in relationships with verbs, which allows the RASP model to get 100% on the ReCOGS test set, and 100% two of the three structural generalizations, and 92% on the obj-pp-to-subj-pp split".
Recall the Wu et al 2023 baseline trained-from-scratch Transformer actually appears to have already partly learned such a non-hierarchical approach as we implement (Figure 3) but not the pp masking (leading to attraction errors not predicted otherwise, Figure 7, Results p6) that would allow it to achieve its full potential. So it is the pp masking largely that prevents us from pursing a learnability argument at this time.

A snippet related to this that I used in one of the internal discussions:
---
We attempt to show in the paper that this non-hierarchical type of solution [argued as in #70 to be have useful generalization properties and layer depth scaling] is relevant to real world Transformers, as the Wu et al 2023 baseline trained-from-scratch Transformer actually appears to have already partly learned such a non-hierarchical approach (Figure 3) but not the pp masking (leading to attraction errors not predicted otherwise, Figure 7, Results p6) that would allow it to achieve its full potential.

[T]o hopefully avoid constructing a feasible but unlearnable model, all the RASP program rules were associated with examples in the official non-augmented training data (Table 2, though we claim feasibility but cannot yet claim overall learnability of the entire solution including the pp/cp masking in this paper), even for example the pp masking is one way to handle the sparse cases of center embedded prepositional phrase modification in the training data (without pp masking our RASP training score would drop). For example, masking the center-embedded pp ("on a tree") in COGS/ReCOGS training sentence “Isabella forwarded a box on a tree to Emma" transforming the sentence to “Isabella forwarded a box _ to Emma” allows us to have noun-verb relationships extracted the same as other non-pp modified v_dat_p1 examples in Table 2 by simply mapping the 3 nouns (Isabella, box, Emma) to agent, theme, recipient in order (same for all v_dat_p1). Another way since the center embedded pp examples are rare in training and the same flat pattern strategy is unaffected by tail pp modification (Figure 7) is just to ignore the rare center embedded pp training examples and not learn masking, which will result in awful generalization set performance (especially on obj-pp-to-subj-pp split which the Wu et al 2023 baseline Transformer gets about 80% wrong on ReCOGS_pos). We already note in our conclusion that we don’t yet know how to get the Transformer trained from scratch to learn the masking rules but we speculate they could be learned from these center embedded prepositional phrase examples in the training data.
---

**Comment by willy-b** [comment]
Note that learnability depends on the architecture, learnability by any neural network is not the same as learnability by a vanilla Transformer. 
For example Rumelhart 1988 found that basic feedforward neural networks could learn by backpropagation the Parity task, but Transformers (which can be handcrafted to compute Parity, see (Chiang and Cholak, 2022) ) are not able to reliably learn Parity (Chiang and Cholak, 2022)  (Deletang et al., 2023). So it is a valid question whether a rule which may be learnable in general in the training set is learnable by a vanilla Transformer specifically.
